src/pkg/shub/client/pull.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package client
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+
+	util "github.com/singularityware/singularity/src/pkg/library/client"
+	"github.com/singularityware/singularity/src/pkg/sylog"
+	pb "gopkg.in/cheggaaa/pb.v1"
+)
+
+// Timeout for an image pull in seconds (2 hours)
+const pullTimeout = 7200
+
+// DownloadImage will retrieve an image from the Container Singularityhub,
+// saving it into the specified file
+func DownloadImage(filePath string, shubRef string, force bool) (err error) {
+	sylog.Debugf("Downloading container from Shub")
+
+	//use custom parser to make sure we have a valid shub URI
+	if ok := isShubPullRef(shubRef); !ok {
+		sylog.Fatalf("Invalid shub URI: %v", err)
+	}
+
+	ShubURI, err := shubParseReference(shubRef)
+	if err != nil {
+		sylog.Fatalf("Failed to parse shub URI: %v", err)
+		return
+	}
+
+	if filePath == "" {
+		filePath = fmt.Sprintf("%s_%s.simg", ShubURI.container, ShubURI.tag)
+		sylog.Infof("Download filename not provided. Downloading to: %s\n", filePath)
+	}
+
+	if !force {
+		if _, err := os.Stat(filePath); err == nil {
+			return fmt.Errorf("image file already exists - will not overwrite")
+		}
+	}
+
+	// Get the image manifest
+	manifest, err := getManifest(ShubURI)
+	if err != nil {
+		sylog.Fatalf("Failed to get manifest from Shub: %v", err)
+		return
+	}
+
+	// Get the image based on the manifest
+	resp, err := http.Get(manifest.Image)
```

Comment:
We should probably use the user-agent for this request too? It is being used when getting the manifest.